### ISSUE TITLE
Run multiple commands with single hook

### ISSUE THREAD

**Issue by rafaelrc7** [neomutt/neomutt#4184]
Hello, I'm trying to run multiple commands with a single hook. This sounds simple enough, but I was not able to do it nor find any information about online.

My ideia is to use `shutdown-hook` to first save changes with `<sync-mail>` and then run a shell script that synchronises the maildir itself. I am able to easily do each alone, writing either `shutdown-hook 'exec sync-mail'` or ``shutdown-hook 'echo `sync-script.sh`'``. However, I was not able to run both, first `sync-mail` and then the script.

Is there a way that I could do that? Thanks.

**Comment by rafaelrc7** [comment]
Ok. I think I did manage to run both commands, by either doing:

``shutdown-hook 'exec sync-mailbox `sync-script.sh`'``

or

``shutdown-hook 'exec sync-mailbox ; echo `sync-script.sh`'``

However it still seems to not be working as expected... The sync-script (that is just a call to mbsync) detects no changes, so read emails are not updated, as if `sync-mailbox` had not been executed. However, if I run the hook two times, by closing, opening and closing it works, so the command is working but the mailbox is only being saved after the program exits. 

How could I fix it?

**Comment by dorsiflexion** [comment]
In fact it seems that `shutdown-hook 'exec sync-mailbox'` is doing exactly nothing.

I'm trying to do something very similar. It doesn't even seem to be a race condition, because when doing
```
shutdown-hook 'exec sync-mailbox; echo `sleep 5`; echo `script.sh`'
```
it is still *not* updated when the script runs.

I'd also greatly appreciate some advice on what's wrong here.

**Comment by rafaelrc7** [comment]
> In fact it seems that `shutdown-hook 'exec sync-mailbox'` is doing exactly nothing.
> 
> I'm trying to do something very similar. It doesn't even seem to be a race condition, because when doing
> ```
> shutdown-hook 'exec sync-mailbox; echo `sleep 5`; echo `script.sh`'
> ```
> it is still *not* updated when the script runs.
> 
> Btw. neomutt updates mailboxes when it quits. So if you are closing, opening, closing and it works, that is no indication that the hook actually did something.
> 
> I'd also greatly appreciate some advice on what's wrong here.

Yes, I also tried using `sleep` in my hook to test if it was some race condition, and had the same results.

I ended up rebinding the exit key (`q`) to run the sync commands and then quit. It works. But it still seems to be something wrong with the `shutdown-hook`.

**Comment by roccoblues** [comment]
Hi,

to clarify: the `shutdown-hook` is executed right? The problem is that `sync-mailbox` doesn't work as expected in it?

My wild guess is that `sync-mailbox` works on the current mailbox and by the time the `shutdown-hook` is executed there's no "current mailbox" anymore. But we need @flatcap expertise here.

**Comment by roccoblues** [comment]
I had a deeper look and the problem doesn't seem to be related to `sync-mailbox`. It looks like be that **no** `exec` command gets executed. With the following `shutdown-hook`

```
shutdown-hook "echo 'before'; exec sync-mailbox; echo 'after'"
```

and some additional logging I get:

```
[2024-03-18 08:44:11]<1> dlg_index() Got op OP_QUIT (198)
[2024-03-18 08:44:11]<5> mutt_startup_shutdown_hook() shutdown
[2024-03-18 08:44:11]<5> mutt_startup_shutdown_hook() hook: echo 'before'; exec sync-mailbox; echo 'after'
[2024-03-18 08:44:11]<5> parse_rc_buffer() buffer:
[2024-03-18 08:44:11]<1> parse_rc_buffer() NT_COMMAND: echo
[2024-03-18 08:44:11]<M> parse_echo() before
[2024-03-18 08:44:11]<5> msgwin_set_text() MW SET: 6, before
[2024-03-18 08:44:11]<5> msgwin_calc_rows() msgwin_calc_rows() => 1
[2024-03-18 08:44:11]<5> msgwin_recalc() recalc done, request WA_REPAINT
[2024-03-18 08:44:11]<5> msgwin_repaint() msgwin repaint done
[2024-03-18 08:44:11]<5> parse_rc_buffer() rc: 0
[2024-03-18 08:44:11]<N> notify_send() sending: NT_COMMAND/13
[2024-03-18 08:44:11]<N> send() send: 7, 0x12001f800
[2024-03-18 08:44:11]<1> parse_rc_buffer() NT_COMMAND: exec
[2024-03-18 08:44:11]<5> mutt_parse_exec() exec: echo 'before'; exec sync-mailbox; echo 'after'
[2024-03-18 08:44:11]<5> mutt_parse_exec() function: sync-mailbox
[2024-03-18 08:44:11]<5> mutt_push_macro_event() ch: 0 op: 174
[2024-03-18 08:44:11]<5> parse_rc_buffer() rc: 0
[2024-03-18 08:44:11]<N> notify_send() sending: NT_COMMAND/14
[2024-03-18 08:44:11]<N> send() send: 7, 0x12001f800
[2024-03-18 08:44:11]<1> parse_rc_buffer() NT_COMMAND: echo
[2024-03-18 08:44:11]<M> parse_echo() after
[2024-03-18 08:44:11]<5> msgwin_set_text() MW SET: 5, after
```

Notice how the `exec sync-mailbox` gets parsed correct but never executed. It's gets pushed as an "event" in `mutt_push_macro_event()` but that's it. I think that is because NeoMutt is no longer in the main loop and nothing is actually handling those events.

I'll see if I can find some changes related to that next.

**Comment by roccoblues** [comment]
Quick update here: we have a, not to ugly, [PR](https://github.com/neomutt/neomutt/pull/4200) that fixes the issue that `exec` functions aren't triggered in the hook.

However the behavior is confusing because the command queue is only triggered after the complete hook is parsed. Fixing this would be a way bigger change.

I suggest something like `alias nm='neomutt && sync-script.sh'` as a temporary workaround for the original issue.
